### PR TITLE
Upgrade Tokio to v1

### LINKED ISSUE #420: Upgrade to Tokio 1.0
rust-opentelemetry currently depends on Tokio 0.2. [Tokio 1.0 released on December 22](https://tokio.rs/blog/2020-12-tokio-1-0) and some of the ecosystem has started moving over to it ([hyper](https://github.com/hyperium/hyper/blob/master/CHANGELOG.md#v0140-2020-12-23), [reqwest](https://github.com/seanmonstar/reqwest/blob/HEAD/CHANGELOG.md#v0110), [rusoto](https://github.com/rusoto/rusoto/blob/master/CHANGELOG.md#0460---2021-01-05), etc.).

### PR DESCRIPTION
**Pull request by mraerino**
*Fixes #420*

I also upgraded these dependencies:
- reqwest
- hyper
- tonic
- prost

### CODE REVIEW

**Comment by mraerino** [comment]
the minimum support rust version for tokio is 1.45: https://github.com/tokio-rs/tokio#supported-rust-versions

do you think we could raise the version here as well? 1.42 seems pretty old

**Comment by TommyCpp** [comment]
> do you think we could raise the version here as well? 1.42 seems pretty old

I think we can. We promised three versions before current stable version, which is 1.49. So we can bump our minimum supported version to 1.45 or 1.46

**Comment by mraerino** [comment]
This is ready for a review now.
I expect this to need a major version bump since a feature flag changed and people can't update automatically via semver since they need to upgrade their tokio version too

**Comment by jtescher** [comment]
@mraerino thanks for working on this, as far as MSRV is concerned we could go either way, for example the `async-std` feature currently doesn't support 1.42 either. Maybe we should just remove the tokio support from the "supported" features for 1.42 and the associated ci workflow and bump it in a subsequent PR? Nice to support folks on older versions in the more stable core even if opt-in features can't support them.

Also pre-1.0 minor changes often include breaking changes, so not as much of a problem there until we hit GA

**Comment by mraerino** [comment]
> Maybe we should just remove the tokio support from the "supported" features for 1.42 and the associated ci workflow and bump it in a subsequent PR?

tokio is used in the tests even if zero features are enabled. i don't think what you propose is possible with the current CI flow

**Comment by djc** [comment]
> Maybe we should just remove the tokio support from the "supported" features for 1.42 and the associated ci workflow and bump it in a subsequent PR? Nice to support folks on older versions in the more stable core even if opt-in features can't support them.

Alternatives:

* Remove tokio (and probably also async-std) from the core crate (as discussed in #403)
* Keep an opentelemetry branch alive for a bit longer with 0.11.x maintenance releases

The reason I hadn't done the upgrade here is there are also some tonic dependencies, and tonic support for tokio 1 has not yet been released. So if we want to prevent spawning multiple runtimes in the background, this PR should (a) wait until tonic 0.4 has been released -- judging from chat activity, this should come pretty soon, or (b) change crates depending on tonic to depend on a released version of opentelemetry, sticking to tokio 0.2 for now.

**Comment by mraerino** [comment]
thanks for pointing out tonic! totally did not see that.
the PR is merged so once they release it we can move forward: https://github.com/hyperium/tonic/pull/530

**Comment by mraerino** [comment]
i made https://github.com/open-telemetry/opentelemetry-rust/pull/424 to test the otlp exporter closer to reality. With that test, CI would be red right now since tonic fails to run in a Tokio v1 executor.

**Comment by djc** [comment]
FYI, tonic 0.4 has been released.

**Comment by mraerino** [comment]
yup, i saw. merging master and upgrading tonic

**Comment by mraerino** [comment]
can you retry the CI workflow? it's a ratelimit failure.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -48,11 +48,11 @@ jobs:
     - uses: actions-rs/toolchain@v1
       with:
         profile: minimal
-        toolchain: 1.42.0
+        toolchain: 1.46.0
         override: true
     - name: Run tests
       run: cargo --version &&
-        cargo test --verbose --manifest-path=opentelemetry/Cargo.toml --features trace,metrics,serialize,tokio,serde,testing &&
+        cargo test --verbose --manifest-path=opentelemetry/Cargo.toml --features trace,metrics,serialize,tokio-support,tokio,serde,testing &&
```
Comment (djc) [inline .github/workflows/ci.yml:55]: `tokio` is implied by `tokio-support`, so we don't need both here:

```suggestion
        cargo test --verbose --manifest-path=opentelemetry/Cargo.toml --features trace,metrics,serialize,tokio-support,serde,testing &&
```

**Review by djc — COMMENTED** [review]
Other than the nit below, this is looking good to me!

**Review by djc — APPROVED** [review]

**Comment by mraerino** [comment]
@djc @jtescher is there a timeline for releasing this?

**Comment by jtescher** [comment]
ASAP

**Pull request merged**